compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/AIETargetDirect.cpp

```diff
@@ -0,0 +1,417 @@
+    // error out if we think the name will most likely be too long
+    // for the artifact generation to succeed. We set this cut-off at 50
+    // characters.
+    if (entryPointName.size() > 50)
+      return exportOp.emitError()
+             << "entry point name: " << entryPointName << "is too long!";
+  }
+  uint64_t ordinalCount = entryPointOrdinals.size();
+
+  std::vector<uint32_t> npuInstrs;
+  std::unique_ptr<llvm::MemoryBuffer> xclbinIn;
+
+  FlatbufferBuilder builder;
+  iree_amd_aie_hal_xrt_ExecutableDef_start_as_root(builder);
+  SmallVector<iree_amd_aie_hal_xrt_XclbinDef_ref_t> xclbinRefs;
+  SmallVector<iree_amd_aie_hal_xrt_AsmInstDef_ref_t> asmInstrRefs;
+
+  // Per entry-point data.
+  // Note that the following vectors should all be of the same size and
+  // element at index #i is for entry point with ordinal #i!
+  SmallVector<std::string> entryPointNamesFb(ordinalCount);
+  SmallVector<uint32_t> xclbinIndices(ordinalCount);
+  SmallVector<uint32_t> asmInstrIndices(ordinalCount);
+
+  for (size_t i = 0; i < entryPointNames.size(); i++) {
+    uint64_t ordinal = entryPointOrdinals.at(entryPointNames[i]);
+
+    entryPointNamesFb[ordinal] = entryPointNames[i];
+
+    SmallString<128> inputMlirPath(workDir);
+    llvm::sys::path::append(inputMlirPath,
+                            entryPointNamesFb[ordinal] + ".aiecc.mlir");
+
+    std::string errorMessage;
+    {
+      auto inputMlirOut = openOutputFile(inputMlirPath, &errorMessage);
+      if (!inputMlirOut) {
+        return moduleOp.emitOpError()
+               << "Failed to write MLIR: " << errorMessage;
+      }
+      deviceOps[i].print(inputMlirOut->os(), OpPrintingFlags().useLocalScope());
+      inputMlirOut->keep();
+    }
+    // we add the entry point to the working directory for xclbin artifacts if
+    // there are multiple entry points so that we dont overwrite the xclbinutil
+    // generated artifacts e.g kernels.json, for different entry points which
+    // will have the same exact names.
+    SmallString<128> entryPointWorkDir(workDir);
+    if (ordinalCount > 1)
+      llvm::sys::path::append(entryPointWorkDir, entryPointNamesFb[ordinal]);
+    SmallString<128> xclbinPath(entryPointWorkDir);
+    llvm::sys::path::append(xclbinPath, entryPointNamesFb[ordinal] + ".xclbin");
+    SmallString<128> npuInstPath(entryPointWorkDir);
+    llvm::sys::path::append(npuInstPath,
+                            entryPointNamesFb[ordinal] + ".npu.txt");
+
+    xilinx::XCLBinGenConfig TK;
+    TK.DisableThreading = options.aie2xclbinDisableTheading;
+    TK.PrintIRAfterAll = options.aie2xclbinPrintIrAfterAll;
+    TK.PrintIRBeforeAll = options.aie2xclbinPrintIrBeforeAll;
+    TK.PrintIRModuleScope = options.aie2xclbinPrintIrModuleScope;
+    TK.TargetArch = "AIE2";
+    TK.TempDir = entryPointWorkDir.str();
+    TK.UseChess = options.useChess;
+    TK.Verbose = options.showInvokedCommands;
+    TK.XCLBinInstanceName = "FOO";
+    TK.XCLBinKernelID = "0x101";
```

Comment:
can we stringify the ordinal into a hex string here? (It will help when we start merging kernels)
here is what I was thinking, Something like this has worked for me.

```

    // Convert ordinal to hexadecimal string for xclbin kernel id.
    std::stringstream ss;
    ss << "0x" << std::hex << ordinal;
    std::string ordinalHex = ss.str();
```